src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const currentState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      for (const key in currentState) {
+        delete currentState[key];
+      }
+      stateHistory.push({ ...currentState });
+    }
+
+    if (action.type === 'addProperties') {
+      Object.assign(currentState, action.extraData);
+      stateHistory.push({ ...currentState });
+    }
+
+    if (action.type === 'removeProperties') {
```

Comment:
Consider adding a check to ensure that `action.keysToRemove` is defined and is an array before iterating over it. This will prevent potential runtime errors if `keysToRemove` is missing or incorrectly formatted in the action object.